### PR TITLE
Add streamerFile-based unitTest for BeamSpotOnlineHLT DQM client

### PR DESCRIPTION
**Pull request by francescobrivio**
#### PR description:
This PR adds the unitTest for the DQM BeamSpotHLT client (`beamhlt_dqm_sourceclient-live_cfg.py`).
Since the client requires to be run on a streamer file (in production it runs on an _ad hoc_ stream `streamDQMOnlineBeamspot`),
I had to make some changes:
 - Added a new `data` directory in `DQM/Integration/data` to store the streamer files
 - Added three `.dat` and three `.jsn` files which are the actual input streamer files for the unitTest
 - Modified the `beamhlt` client so that when `unitTest=True` the standard DAS source is owerwritten with 
   `process.source = cms.Source("DQMStreamerReader"...`

@cms-sw/dqm-l2 @jfernan2 please let me know if this approach is acceptable for DQM.
Another possible approach would be to introduce a new config file in `DQM/Integration/python/config/`, something like `streamerinputsource_cfi.py`, but **at the moment** this is the only client with a unitTest that requires a streamer input file as the other two are commented out, see https://github.com/cms-sw/cmssw/blob/f3a3808f4a0651fed52cc72efc9c5d4dd3603e62/DQM/Integration/test/BuildFile.xml#L30-L33

#### PR validation:
`scram b runtests` runs and the new unitTest runs succesfully.

**!!** PLEASE NOTE that this PR only changes the unitTest behavior, **NOT** the production one **!!**

#### Backport:
Not a backport, but a 12_3_X backport will be provided as soon as I get feedback on this PR from @cms-sw/dqm-l2.

FYI @dzuolo @gennai

### CODE REVIEW

**Comment by francescobrivio** [comment]
@cmsbuild please test

**Comment by francescobrivio** [comment]
The new unittest runs successfully: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e430a3/24166/unitTests/src/DQM/Integration/test/TestDQMOnlineClient-beamhlt_dqm_sourceclient/testing.log

@cms-sw/dqm-l2 please let me know if the proposed strategy described in the PR description is ok for you.

**Comment by mmusich** [comment]
Just to be sure,  

> Added a new data directory in DQM/Integration/data to store the streamer files

shouldn't these go to `cms-data` instead?

**Comment by francescobrivio** [comment]
> shouldn't these go to `cms-data` instead?

Yea...this was why I wanted an advice from @cms-sw/dqm-l2 on how to better proceed: maybe a directory in `cms-data` already exists for DQM (at which point one could also put there the streamer files for the other 2 clients that require `streamDQMCalibration` as I wrote in the description).

**Comment by mmusich** [comment]
> maybe a directory in cms-data already exists for DQM 

As far as I can tell, not for this package.

SmartSelect_20220425-095823_Chrome

**Comment by jfernan2** [comment]
@francescobrivio I do not want to interfere with DQM convener's opinion, so I let @emanueleusai in charge to decide, but if you want my opinion, I believe @mmusich is right and ORP managers would prefer to use cms-data instead, even if this is affecting Online DQM mainly: CMSSW package should contain code only, so at least the .dat and .jsn files should go there.

**Comment by perrotta** [comment]
I agree that a DQM-Integration (or whatever else name is deemed appropriate) direcotory in cms-data could be a preferable option for storing these files.
Normally @smuzaffar can take care of providing that area, if you also agree with it

**Comment by qliphy** [comment]
@francescobrivio   The directory is ready https://github.com/cms-data/DQM-Integration

**Comment by smuzaffar** [comment]
thanks @qliphy for taking care of the data repository.

**Comment by francescobrivio** [comment]
@qliphy @smuzaffar could you suggest how to update the path in:
```
process.source = cms.Source("DQMStreamerReader",
   runInputDir = cms.untracked.string('src/DQM/Integration/data'),
)
```
to read the files from cms-data (https://github.com/cms-data/DQM-Integration/pull/1)?

**Comment by qliphy** [comment]
> @qliphy @smuzaffar could you suggest how to update the path in:
> 
> ```
> process.source = cms.Source("DQMStreamerReader",
>    runInputDir = cms.untracked.string('src/DQM/Integration/data'),
> )
> ```
> 
> to read the files from cms-data ([cms-data/DQM-Integration#1](https://github.com/cms-data/DQM-Integration/pull/1))?

You may refer to 
https://github.com/cms-sw/cmssw/blob/master/L1Trigger/L1TMuonCPPF/python/emulatorCppfDigis_cfi.py#L12
with files from https://github.com/cms-data/L1Trigger-L1TMuon

**Comment by smuzaffar** [comment]
@francescobrivio , as [runInputDir](https://github.com/cms-sw/cmssw/blob/7d5358ff0e31e78fbcb737c05d218a55f5d82104/DQMServices/StreamerIO/plugins/DQMStreamerReader.cc#L28) is std::string and not a `FileInPath` type , so I think you can do something like this
```
import os
dqm_integration_data = [os.path.join(dir,'DQM/Inetgration/data') for dir in os.getenv('CMSSW_SEARCH_PATH','').split(":") if os.path.exists(os.path.join(dir,'DQM/Inetgration/data'))][0]
process.source = cms.Source("DQMStreamerReader",
   runInputDir = cms.untracked.string(dqm_integration_data),
)
```

**Comment by francescobrivio** [comment]
test parameters:

 - pull_request = https://github.com/cms-data/DQM-Integration/pull/1

**Comment by francescobrivio** [comment]
>     * Found files with invalid states:
>       
>       * DQM/Integration/data/run346373/run346373_ls0027_streamDQMOnlineBeamspot_sm-c2a11-43-01.dat:
>         
>         * Added: [15f3590](https://github.com/cms-sw/cmssw/commit/15f359021bfc33c8390c0ebdc8c6c0d92ec14917)
>         * Deleted: [de608ac](https://github.com/cms-sw/cmssw/commit/de608acc238cf31eca09612ec98b35c9f854be6c)
>       * DQM/Integration/data/run346373/run346373_ls0028_streamDQMOnlineBeamspot_sm-c2a11-43-01.jsn:
>         
>         * Added: [15f3590](https://github.com/cms-sw/cmssw/commit/15f359021bfc33c8390c0ebdc8c6c0d92ec14917)
>         * Deleted: [de608ac](https://github.com/cms-sw/cmssw/commit/de608acc238cf31eca09612ec98b35c9f854be6c)
>       * DQM/Integration/data/run346373/run346373_ls0027_streamDQMOnlineBeamspot_sm-c2a11-43-01.jsn:
>         
>         * Added: [15f3590](https://github.com/cms-sw/cmssw/commit/15f359021bfc33c8390c0ebdc8c6c0d92ec14917)
>         * Deleted: [de608ac](https://github.com/cms-sw/cmssw/commit/de608acc238cf31eca09612ec98b35c9f854be6c)
>       * DQM/Integration/data/run346373/run346373_ls0028_streamDQMOnlineBeamspot_sm-c2a11-43-01.dat:
>         
>         * Added: [15f3590](https://github.com/cms-sw/cmssw/commit/15f359021bfc33c8390c0ebdc8c6c0d92ec14917)
>         * Deleted: [de608ac](https://github.com/cms-sw/cmssw/commit/de608acc238cf31eca09612ec98b35c9f854be6c)
>       * DQM/Integration/data/run346373/run346373_ls0029_streamDQMOnlineBeamspot_sm-c2a11-43-01.jsn:
>         
>         * Added: [15f3590](https://github.com/cms-sw/cmssw/commit/15f359021bfc33c8390c0ebdc8c6c0d92ec14917)
>         * Deleted: [de608ac](https://github.com/cms-sw/cmssw/commit/de608acc238cf31eca09612ec98b35c9f854be6c)
>       * DQM/Integration/data/run346373/run346373_ls0029_streamDQMOnlineBeamspot_sm-c2a11-43-01.dat:
>         
>         * Added: [15f3590](https://github.com/cms-sw/cmssw/commit/15f359021bfc33c8390c0ebdc8c6c0d92ec14917)
>         * Deleted: [de608ac](https://github.com/cms-sw/cmssw/commit/de608acc238cf31eca09612ec98b35c9f854be6c)

I'll squash and force-push in order to get rid of these warnings.

**Comment by francescobrivio** [comment]
@cmsbuild please test

**Comment by francescobrivio** [comment]
the [beamhlt unitTest](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e430a3/24224/unitTests/src/DQM/Integration/test/TestDQMOnlineClient-beamhlt_dqm_sourceclient/testing.log) finishes successfully, but the correct path is not picked up:
```
26-Apr-2022 14:27:07 CEST  Directory does not exist: /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_4_X_2022-04-26-1100/src/DQM/Integration/run346373
26-Apr-2022 14:27:07 CEST  Streamer state changed: 0->2
26-Apr-2022 14:27:07 CEST  Waiting for the first lumi in order to initialize.
26-Apr-2022 14:27:07 CEST  End of run reached before DQMStreamerReader was initialised.
```

@smuzaffar is the path to the directory resolved correctly?

**Comment by smuzaffar** [comment]
@francescobrivio , you have missed `/data`
```
dqm_integration_data = [os.path.join(dir,'DQM/Inetgration/data') for dir in os.getenv('CMSSW_SEARCH_PATH','').split(":") if os.path.exists(os.path.join(dir,'DQM/Inetgration/data'))][0]
```

File: DQM/Integration/python/clients/beamhlt_dqm_sourceclient-live_cfg.py [hunk 1]
```diff
@@ -48,6 +48,30 @@
 if unitTest:
   process.load("DQM.Integration.config.unittestinputsource_cfi")
   from DQM.Integration.config.unittestinputsource_cfi import options
+
+  # Overwrite source of the unitTest to use a streamer file instead of the DAS query output
+  print("[beamhlt_dqm_sourceclient-live_cfg]:: Overriding DAS input to use a streamer file")
+
+  # Read streamer files from https://github.com/cms-data/DQM-Integration
+  import os
+  dqm_integration_data = [os.path.join(dir,'DQM/Integration') for dir in os.getenv('CMSSW_SEARCH_PATH','').split(":") if os.path.exists(os.path.join(dir,'DQM/Integration'))][0]
```
Comment (smuzaffar) [inline DQM/Integration/python/clients/beamhlt_dqm_sourceclient-live_cfg.py:57]: @francescobrivio , please use `DQM/Integration/data`

Comment (francescobrivio) [reply DQM/Integration/python/clients/beamhlt_dqm_sourceclient-live_cfg.py:57]: `data` was originally added by me beacuse I manually added a `data` directory inside cmssw/DQM/Integration, the new `cms-data` repo does not have the `data` subdir.
Will the fileInPath automatically resolve it to the correct cms-data repo?

Comment (smuzaffar) [reply DQM/Integration/python/clients/beamhlt_dqm_sourceclient-live_cfg.py:57]: yes the `cms-data/SubSystem-Package` contents are installed as `somepath/SubSystem/Package/data` (where `<somepath>` is added to `CMSSW_SEARCH_PATH` env . This is done to match the `cmssw/src/SubSystem/Package/data` .

Comment (smuzaffar) [reply DQM/Integration/python/clients/beamhlt_dqm_sourceclient-live_cfg.py:57]: For `cms-data/DQM-Integration`, the install path has
```
> ls /cvmfs/cms-ci.cern.ch/week0/PR_b061bf72/slc7_amd64_gcc10/cms/data-DQM-Integration/V00-00-00/DQM/Integration/data 
README.md  run346373
```

Comment (francescobrivio) [reply DQM/Integration/python/clients/beamhlt_dqm_sourceclient-live_cfg.py:57]: @smuzaffar thank you very much for the explanation! 
I'll fix the path in the next commit.

Comment (francescobrivio) [reply DQM/Integration/python/clients/beamhlt_dqm_sourceclient-live_cfg.py]: addessed in [d310c28](https://github.com/cms-sw/cmssw/pull/37667/commits/d310c2820b6de1fb7017e5c34064a0834b6673c4)

**Comment by francescobrivio** [comment]
@cmsbuild please test

 - third time's the charm

**Comment by francescobrivio** [comment]
The [unitTest](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e430a3/24227/unitTests/src/DQM/Integration/test/TestDQMOnlineClient-beamhlt_dqm_sourceclient/testing.log) now runs succesfully! Thanks @smuzaffar for the help!

I'll open a backport to 12_3_X soon.

**Comment by emanueleusai** [comment]
+1

**Comment by qliphy** [comment]
+1

**Pull request merged**